src/util/newconfig/ConfigDefinition.cpp

```diff
@@ -0,0 +1,174 @@
+     {"log_rotation_size", ConfigValue{ConfigType::Integer}.defaultValue(2048)},
+     {"log_directory_max_size", ConfigValue{ConfigType::Integer}.defaultValue(50 * 1024)},
+     {"log_rotation_hour_interval", ConfigValue{ConfigType::Integer}.defaultValue(12)},
+     {"log_tag_style", ConfigValue{ConfigType::String}.defaultValue("uint")},
+     {"extractor_threads", ConfigValue{ConfigType::Integer}.defaultValue(2u)},
+     {"read_only", ConfigValue{ConfigType::Boolean}.defaultValue(false)},
+     {"txn_threshold", ConfigValue{ConfigType::Integer}.defaultValue(0)},
+     {"start_sequence", ConfigValue{ConfigType::String}.optional()},
+     {"finish_sequence", ConfigValue{ConfigType::String}.optional()},
+     {"ssl_cert_file", ConfigValue{ConfigType::String}.optional()},
+     {"ssl_key_file", ConfigValue{ConfigType::String}.optional()},
+     {"api_version.min", ConfigValue{ConfigType::Integer}},
+     {"api_version.max", ConfigValue{ConfigType::Integer}}}
+};
+
+ObjectView
+ClioConfigDefinition::getObject(std::string_view prefix, std::optional<std::size_t> idx) const
+{
+    auto const prefixWithDot = std::string(prefix) + ".";
+    for (auto const& [mapKey, mapVal] : map_) {
+        if (mapKey.starts_with(prefixWithDot) && (std::holds_alternative<ConfigValue>(mapVal) || !idx.has_value()))
+            return ObjectView{prefix, *this};
+
+        if (mapKey.starts_with(prefixWithDot) && std::holds_alternative<Array>(mapVal)) {
+            ASSERT(std::get<Array>(mapVal).size() > idx, "index provided is out of scope");
+            return ObjectView{prefix, idx.value(), *this};
+        }
+    }
+    ASSERT(false, "Key {} is not found in config", prefixWithDot);
+    return ObjectView{"", *this};
+}
+
+ArrayView
+ClioConfigDefinition::getArray(std::string_view prefix) const
+{
+    auto key = std::string(prefix);
+    if (!prefix.contains(".[]"))
+        key += ".[]";
+
+    for (auto const& [mapKey, mapVal] : map_) {
+        if (mapKey.starts_with(key)) {
+            ASSERT(std::holds_alternative<Array>(mapVal), "Trying to retrieve an Array when value is an object ");
+            return ArrayView{key, *this};
+        }
+    }
+    ASSERT(false, "Key {} is not found in config", key);
+    return ArrayView{"", *this};
+}
+
+ValueView
+ClioConfigDefinition::getValue(std::string_view fullKey) const
+{
+    ASSERT(map_.contains(fullKey), "key {} does not exist in config", fullKey);
+    if (std::holds_alternative<ConfigValue>(map_.at(fullKey))) {
+        return ValueView{std::get<ConfigValue>(map_.at(fullKey))};
+    }
+    ASSERT(false, "Value of key {} is an Array, not an object", fullKey);
+    return ValueView{ConfigValue{}};
+}
+
+Array const&
+ClioConfigDefinition::atArray(std::string_view key) const
+{
+    ASSERT(map_.contains(key), "Current string {} is a prefix, not a key of config", key);
+    ASSERT(std::holds_alternative<Array>(map_.at(key)), "Value of {} is not an array", key);
```

Comment:
Better to get an iterator first and then use it. It will save two searches in the map.